### ISSUE TITLE
[JENKINS-9913] Not obvious why some post-build tasks enforce serial behavior even when builds are concurrent

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-06-a2#10992]
<p>We're experiencing an issue with concurrent builds where Jenkins appears to be associating separate builds (run on different machines) such that they won't be marked as completed until all jobs are completed.  For example, if we kick off 5 concurrent builds on 5 different nodes, builds 1-4 won't be marked as completed if build #5 is still running, even though builds 1-4 are finished.  I've seen a report of someone experiencing this issue elsewhere:</p>

<p><a href="http://groups.google.com/group/jenkinsci-users/browse_thread/thread/e477e25910266d2a?fwc=1" class="external-link" target="_blank" rel="nofollow noopener">http://groups.google.com/group/jenkinsci-users/browse_thread/thread/e477e25910266d2a?fwc=1</a></p>

<p>but a solution wasn't posted.  We do not have the batch plugin or the locks and latches plugin installed.  We've disabled all post-build processing and switched between different containers (Glassfish/Tomcat), but the problem persists.  I couldn't find an issue logged for this other than the aforementioned posting.</p>

**Attachments:**

- [jenkins1.jpeg](https://issues.jenkins.io/secure/attachment/22736/jenkins1.jpeg)
- [jenkins2.jpeg](https://issues.jenkins.io/secure/attachment/22735/jenkins2.jpeg)

---
<details><summary><i>Originally reported by <a title="Philip Metting van Rijn" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=pomvr">pomvr</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-9913" target="_blank">Not obvious why some post-build tasks enforce serial behavior even when builds are concurrent</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Jesse Glick" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>
<li><b>reported by</b>: pomvr
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2013-08-06T22:09:21+00:00
<li><b>votes</b>: 19
<li><b>watchers</b>: 34
<li><b>imported</b>: 2025-11-06
</ul></i>
</details>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svs57">svs57</a>:</i>
<p>I am afraid that no one is working on a the investigation of this bug.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=xavier_leprevost">xavier_leprevost</a>:</i>
<p>It would be nice to know if someone will investigate this problem.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kutzi">kutzi</a>:</i>
<p>Similar to <a href="https://issues.jenkins.io/browse/JENKINS-10234" title="Junit result archiver getting stuck for a long time in concurrent builds" class="issue-link" data-issue-key="JENKINS-10234"><del>JENKINS-10234</del></a> - for some causes even the same.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kutzi">kutzi</a>:</i>
<p>There are several cause described here. Some are JUnit archiving, which is specifically handled in <a href="https://issues.jenkins.io/browse/JENKINS-10234" title="Junit result archiver getting stuck for a long time in concurrent builds" class="issue-link" data-issue-key="JENKINS-10234"><del>JENKINS-10234</del></a>. Some seem to be related to email notification.<br/>
I dare to say that in almost all cases this is a feature and not a bug as the logic for email notification and JUnit result archiving needs to wait until the previous builds are finished.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svs57">svs57</a>:</i>
<p>It is not a feature. All concurent builds execute in separated workspace and must  archive its artifacts independently. No need to wait others builds if current have finished.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kutzi">kutzi</a>:</i>
<p>Sergey, do you mean 'artifact archiving' or 'JUnit result archiving'. In the former case, you would be probably right, but I've seen no comment here that artifact archiving is blocking, too.<br/>
In the case of JUnit: yes it MUST block to calculate the diff (regressions et.al.) to the previous test results. So it is a feature.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svs57">svs57</a>:</i>
<p>I mean all result of concurent job.<br/>
For example. I have Jenkins job that execute E2E tests on different QA servers. I started 10 jobs. One job running 5 hours. But others takes 1 hour. Then I can't see results of finished job and have to wait 4 hours to see it. We must have instrument to prevent this.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svs57">svs57</a>:</i>
<p>Do anybody know how to solve this problem? Any patch, workaround...?<br/>
I can't remove post build steps for the job</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=teh444">teh444</a>:</i>
<p>Happening in 1.515</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svs57">svs57</a>:</i>
<p>>Workaround is to remove all post-Build-Actions<br/>
It is not workaround.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=teh444">teh444</a>:</i>
<p>Sergey, I may not have made myself clear.</p>

<p>For example: <br/>
Project 'A' is the project you wish to run in parallel.</p>

<p>Project 'A' has a bunch of really long shell/bat commands and requires some sort of post-build action that is causing this defect.<br/>
Which is: if Project 'A' is run in parallel and contains post-build actions, it will not release until it completes.</p>

<p>The solution: Create a second project 'B' that contains all your post-build actions from 'A'.<br/>
Finally, remove all the post-build actions in 'A' (as they are all duplicated in project 'B') and have 'A' trigger 'B' as the final step.</p>

<p>Additional notes:<br/>
You will need to use the "NodeLabel Parameter Plugin" to ensure that this project is run on the same machine. <br/>
You may also need to give it the ${WORKSPACE} parameter (and any others) if your post-build actions need to manipulate the artefacts generated from Project 'A'.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=svs57">svs57</a>:</i>
<p>Thank you, 4 4. Now it's clear.<br/>
This will cause new problems, because my tests execute notification job in post-build step. This job analize what was upstream and send mail <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/sad.png" height="16" width="16" align="absmiddle" alt="" border="0"/><br/>
I have to chage logic in notify job.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>Not a bug per se, but Jenkins needs to make sure that post-build actions which require the previous build to be complete (a) are documented to do so, e.g. in inline help; (b) print a helpful message to the build log when waiting for a previous build. And of course wherever feasible, offer an option to not block in this way, or to perform the processing dependent on the previous build asynchronously, e.g. in a <tt>RunListener</tt>.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in jenkins<br/>
User: Jesse Glick<br/>
Path:<br/>
 core/src/main/java/hudson/model/CheckPoint.java<br/>
 core/src/main/java/hudson/model/Run.java<br/>
 core/src/main/java/hudson/tasks/BuildStepMonitor.java<br/>
 core/src/main/java/hudson/tasks/junit/JUnitResultArchiver.java<br/>
 core/src/main/resources/hudson/model/Messages.properties<br/>
<a href="http://jenkins-ci.org/commit/jenkins/eec307511c80112274d27f2a840d9f96cda784d3" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/eec307511c80112274d27f2a840d9f96cda784d3</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED JENKINS-9913&#93;</span> At least print a diagnostic to the build log if we are waiting on a checkpoint.</p>


<p>Compare: <a href="https://github.com/jenkinsci/jenkins/compare/e5f5402cc2fd...eec307511c80" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/compare/e5f5402cc2fd...eec307511c80</a></p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dogfood">dogfood</a>:</i>
<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/blue.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/jenkins_main_trunk/2789/" class="external-link" target="_blank" rel="nofollow noopener">jenkins_main_trunk #2789</a><br/>
     <span class="error">&#91;FIXED JENKINS-9913&#93;</span> At least print a diagnostic to the build log if we are waiting on a checkpoint. (Revision eec307511c80112274d27f2a840d9f96cda784d3)</p>

<p>     Result = SUCCESS<br/>
Jesse Glick : <a href="https://github.com/jenkinsci/jenkins/commit/eec307511c80112274d27f2a840d9f96cda784d3" class="external-link" target="_blank" rel="nofollow noopener">eec307511c80112274d27f2a840d9f96cda784d3</a><br/>
Files : </p>
<ul>
	<li>core/src/main/java/hudson/tasks/junit/JUnitResultArchiver.java</li>
	<li>core/src/main/java/hudson/tasks/BuildStepMonitor.java</li>
	<li>core/src/main/resources/hudson/model/Messages.properties</li>
	<li>core/src/main/java/hudson/model/CheckPoint.java</li>
	<li>core/src/main/java/hudson/model/Run.java</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>@surya548:</p>

<blockquote><p>I am not sure how printing to logs resolves this.</p></blockquote>

<p>As noted above, it just makes it clear why a given plugin is blocking the way it is. If there is a particular plugin which is blocking which you believe should not block, that should be filed in a separate issue.</p>

<blockquote><p>can we make Run.waitForCheckpoint an instance method instead of a static method</p></blockquote>

<p>I see no reason why that would be necessary, but this is probably not the place to discuss it anyway.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=surya548">surya548</a>:</i>

<p>> If there is a particular plugin which is blocking which you believe should not block</p>

<p>I dont want any plugin to block on checkpoint because it doesn't make sense to compare two arbitrary branches in the repo.</p>

<p>> that should be filed in a separate issue. </p>

<p>I have so many plugins  that do this  here are just few example  example : junitarchiver, tap, checkstyle, findbugs , corbertura </p>

<p>I would like a global disable checkpointing option.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<blockquote><p>it doesn't make sense to compare two arbitrary branches in the repo</p></blockquote>

<p>True, but this needs to be solved at a higher level in Jenkins, by creating a separate <tt>AbstractProject</tt> for each branch, so that each has a linear build history. There is some work TBA that accomplishes this.</p>

<blockquote><p>I would like a global disable checkpointing option.</p></blockquote>

<p>Sorry, this needs to be implemented on a per-plugin basis, since plugins are not necessarily written to behave gracefully when their checkpoint expectations are unmet.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=surya548">surya548</a>:</i>
<p>> so that each has a linear build history.<br/>
We dont care about none of this linear history, checkpointing stuff if builds are hung for days deadlocked on checkpoints. <br/>
There should be a way to turn this feature off. I think most people use git with lots of branches anyways these days and none of this is relevant/useful. </p>

<p>>plugins are not necessarily written to behave gracefully when their checkpoint expectations are unmet</p>

<p>if we just noop  Run.waitForCheckpoint based on some global setting wouldn't plugins just assume there is nothing to compare against and behave gracefully? </p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<blockquote><p>I am not sure if it is practical for me to open bugs against 9 plugins</p></blockquote>

<p>Why not?</p>

<blockquote><p>feature designed for SVN/CVS</p></blockquote>

<p>Applies equally to Git as to SVN; both support branches, and for any such SCM it is preferable to have one <tt>AbstractProject</tt> per branch so that each has a sensible linear history (as previously mentioned there is ongoing work in this regard). If you do that, and suppress parallel builds within a branch (i.e. may have concurrent builds only of distinct branches), then there is no further issue.</p>

<p>Now even within a branch it is sometimes desirable to permit parallel builds, when there is a premium on quick feedback over build-to-build comparisons; in that case plugins should be configurable to do no comparison to the previous build (i.e. you are intentionally waiving your right to information such as whether a given test case failure was a regression), and should not use checkpoints either.</p>

<blockquote><p>some optional feature</p></blockquote>

<p>The original behavior of build steps was to be unconditionally serialized. For compatibility, that default must remain. New plugins should consider their actual needs and override the relevant method to specify it (optionally introducing finer-grained checkpoints).</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>Around half of those have more than one commit, as I said.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>I am trying to get a patch through in <a href="https://issues.jenkins.io/browse/JENKINS-16376" title="BuildStepMonitor.BUILD makes concurrent builds wait, could be changed to BuildStepMonitor.NONE?" class="issue-link" data-issue-key="JENKINS-16376"><del>JENKINS-16376</del></a>; please push for changes in other affected plugins.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=byronbrummer">byronbrummer</a>:</i>
<p>I have to wholeheartedly agree with surya et al: There needs to be a (non-plugin) switch for this.  Perhaps not a global switch (although, why not?), but at an absolute minimum a per-job switch.  This is a Jenkins-wide issue and giving plugins enough rope to hang the users does not strike me as the best process.</p>

<p>The only "workaround" really is to throw out the entire Post-Build functionality and recreated it from whole cloth as Build Steps.  That is really what this bug means: The entire Post-Build feature must be considered suspect and unreliable.  At the very minimum there should be a huge user warning around the Concurrent Build check box telling users that Post-Build actions may likely serialize their jobs anyway and should be avoided.</p>

<hr />

<p>It should also be noted that Jenkins jobs are leveraged for MUCH more than simply "pulling code from an SCM and compiling it" type jobs.  There are a huge number of utility tasks around a typical development lifecycle (deployments, environment setups and refreshes, data jobs, etc) that fit very well into Jenkins.  In the real world I see such utility jobs far, far outnumber "real" build jobs.  I mention this because of all the debate above about Git vs Subversion workflows and the like: It doesn't matter...in real world Jenkins usage it's common to find that MOST jobs aren't calling SCM <b>at all</b> because they aren't "buliding" anything, rendering all such workflow debates mute.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>Build step monitors are an aspect of <em>any</em> build step, including both “builders” and “post-build steps” (recorders and notifiers), so that is an artificial distinction. Prior to the introduction of concurrent builds, plugins were free to assume that a build with a lower number than the current one was in fact completed—and some did in fact make that assumption. Monitors were introduced to ensure full binary compatibility. All plugins with build steps need to be updated to specify the monitor style they actually require, which is usually none. This has to be done on a case-by-case basis, while reviewing the plugin code for calls such as <tt>getPreviousBuild</tt> and ensuring that the logic continues to make sense when this build might be finishing well before earlier builds.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=notagain">notagain</a>:</i>
<p>Got bitten by the same thing. Now planning to move the post build stuff into an additional last step of the build job. <br/>
Only problem is that I need to get data from the slave to master before that. And that is also a Post build action (I use Copy-To-Slave plugin). Any way I can do it as a part of build job only?</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=reksiopl">reksiopl</a>:</i>
<p>To resolve this problem use xplugin as an intermediary to nunit/junit.  </p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ericc">ericc</a>:</i>
<p>My apologies if this is the wrong please to ask this but I really did look for the right place and couldn't find it - is there a schedule for when this fix will be released? I am running into this issue and would benefit from the fix.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally related to: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-10234">JENKINS-10234</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally related to: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-21246">JENKINS-21246</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally related to: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-16376">JENKINS-16376</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Originally depends on: <a href="https://github.com/timja-org/jenkins-gh-issues-poc-11-06-a2/issues?q=in%3Atitle%20JENKINS-6854">JENKINS-6854</a>]</i>
